### DISCUSSION TITLE
feat: translate a struct instance into a map

### DISCUSSION THREAD

**Discussion by Hunam6** [vlang/v#15060] [Ideas]
```v
struct ComplexData {
	str   string
	other Other
}

struct Other {
	number int
}

fn main() {
	in_data := ComplexData{}
	out_data := {
		str:   ''
		other: {
			number: 0
		}
	}
}
```
I really think we need a way to be able to transform any struct instance into a map. In the example above `in_data` should be able to be transformed into `out_data`. Where the type of `out_data` could be something like:
```v
type Any = []Any | bool | f32 | f64 | i64 | int | map[string]Any | string | u64
```

There was a small discussion on Discord about it: https://canary.discord.com/channels/592103645835821068/592115457029308427/983073751816085524

**Comment by depperm** [comment]
doesn't the json module do this? Can you provide more of an example of what you are trying to accomplish?

```
import json

struct ComplexData {
	str   string
	other Other
}

struct Other {
	number int
}

fn main() {
    in_data := ComplexData{str: 'test', other: Other{number: 42}}
    out_data := json.decode(ComplexData, json.encode(in_data)) or {
        eprintln('Failed to decode json, error: $err')
        return
    }
    println(out_data)
}
```

**Comment by Hunam6** [comment]
In your example `out_data` isn't a map
```v
println(in_data == out_data) // true
```

**Comment by depperm** [comment]
a struct is basically a map with a definition/defined types, can you explain better what you want?

**Comment by Hunam6** [comment]
`map[string]Any` is the closest match to JSON: string keys, values of any type, you can add and remove keys/values at glance moreover it's safe because of the explicit type cast needed to use any value.

**Comment by Hunam6** [comment]
For the web especially it's good have this flexibility

**Comment by Hunam6** [comment]
My precise use-case is for [Voak](https://github.com/hunam6/voak), I want to implement template engines, as an input I want the use to be able to put anything they want, like a combination of "basic" values, maps, structs etc.

This below is an example on basically how I want it to work
```v
import hunam6.voak

struct MyData {
	a_map   map[string]int
	params  map[string]string
	name    string
	age     int
	numbers []int
	other   OtherData
}

struct OtherData {
	foo string
}

fn main() {
	mut app := voak.App{}
	mut router := voak.Router{}

	router.get('/complex/:name/*path*', fn (mut ctx voak.Ctx) {
		ctx.res.body = voak.tmpl('/assets/index.html', MyData{
			params: ctx.get_query()
			name: 'Hunam'
			age: 17
			numbers: [
				1,
				2,
				3,
			]
			other: OtherData{
				foo: 'bar'
			}
		})
	})

	app.use(router.get_routes)
	app.listen(port: 8080)?
}
```

I'm sure there are other use cases too.
